grpc/federation/cel.go

```diff
@@ -832,3 +859,169 @@ func SetCELValue[T any](ctx context.Context, param *SetCELValueParam[T]) error {
 	}
 	return nil
 }
+
+// nullValueFuncReplacer is a feature that allows to compare typed null and null value correctly.
+// It parses the expression and wraps the message with grpc.federation.cast.null_value if the message is compared to null.
+type nullValueFuncReplacer struct {
+	checkedExpr *exprpb.CheckedExpr
+	lastID      int64
+	replaced    bool
+	unsupported bool
+}
+
+func newNullValueFuncReplacer() *nullValueFuncReplacer {
+	return &nullValueFuncReplacer{}
+}
+
+func (r *nullValueFuncReplacer) init(checkedExpr *exprpb.CheckedExpr) {
+	var lastID int64
+	for k := range checkedExpr.GetReferenceMap() {
+		if lastID < k {
+			lastID = k
+		}
+	}
+	for k := range checkedExpr.GetTypeMap() {
+		if lastID < k {
+			lastID = k
+		}
+	}
+	r.checkedExpr = checkedExpr
+	r.lastID = lastID
+	r.replaced = false
+	r.unsupported = false
+}
+
+func (r *nullValueFuncReplacer) nextID() int64 {
+	r.lastID++
+	return r.lastID
+}
+
+func (r *nullValueFuncReplacer) Replace(checkedExpr *exprpb.CheckedExpr) bool {
+	r.init(checkedExpr)
+	newExprVisitor().Visit(checkedExpr.GetExpr(), func(e *exprpb.Expr) {
+		switch e.GetExprKind().(type) {
+		case *exprpb.Expr_CallExpr:
+			r.replaceCall(e)
+		case *exprpb.Expr_ComprehensionExpr:
+			// Comprehension is not supported by parser.Unparse.
+			r.unsupported = true
```

Comment:
If a comparison with `null` is made while using comprehension, the comparison with `null` will be performed without being replaced. Is it okay not to treat this as an error ?

Reply:
I see. I will try to fix it so that it parses once on the resolver side and returns a validation error if comprehension is used.

Reply:
This exception handling is not needed and I removed it.
Thank you for pointing it out 🙏